The Wine Quality dataset contains data on various physicochemical properties of red and white variants of Portuguese "Vinho Verde" wine. It is commonly used for classification and regression tasks in machine learning. The goal is typically to predict the quality of the wine based on its attributes.

### Dataset Details

- **Source**: The dataset is available from the UCI Machine Learning Repository.
- **Attributes**: It includes 11 input variables and 1 output variable.

### Attributes

1. **Fixed Acidity**: Most acids involved with wine or fixed or nonvolatile (do not evaporate readily).
2. **Volatile Acidity**: The amount of acetic acid in wine, which at too high of levels can lead to an unpleasant, vinegar taste.
3. **Citric Acid**: Found in small quantities, citric acid can add 'freshness' and flavor to wines.
4. **Residual Sugar**: The amount of sugar remaining after fermentation stops, important for determining the sweetness of the wine.
5. **Chlorides**: The amount of salt in the wine.
6. **Free Sulfur Dioxide**: The free form of SO2 exists in equilibrium between molecular SO2 (as a dissolved gas) and bisulfite ion; it prevents microbial growth and the oxidation of wine.
7. **Total Sulfur Dioxide**: Amount of free and bound forms of SO2; in low concentrations, SO2 is mostly undetectable in wine, but at free SO2 concentrations over 50 ppm, SO2 becomes evident in the nose and taste of wine.
8. **Density**: The density of water is close to that of water depending on the percent alcohol and sugar content.
9. **pH**: Describes how acidic or basic the wine is.
10. **Sulphates**: A wine additive which can contribute to sulfur dioxide gas (SO2) levels, acts as an antimicrobial and antioxidant.
11. **Alcohol**: The percent alcohol content of the wine.

### Output Variable

- **Quality**: Wine quality score (between 0 and 10).

### Applications

- **Classification**: Predicting the quality category of the wine (e.g., low, medium, high).
- **Regression**: Predicting the exact quality score of the wine.

### Accessing the Dataset

You can download the Wine Quality dataset from the UCI Machine Learning Repository:
- [Wine Quality Dataset (UCI)](https://archive.ics.uci.edu/ml/datasets/wine+quality)


In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from matplotlib import cm
warnings.filterwarnings('ignore')
pd.options.display.float_format = "{:.2f}".format


In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


##LOAD DATASET

In [ ]:
df_red =pd.read_csv("/content/drive/MyDrive/0.Latest_DS_Course/Statistics/Descriptive/8.EDA/data/winequality-red.csv",delimiter=";")

In [ ]:
df_white =pd.read_csv("/content/drive/MyDrive/0.Latest_DS_Course/Statistics/Descriptive/8.EDA/data/winequality-white.csv",delimiter=";")

##UNDERSTANDNG DATA

#### RED WINE DATASET

In [ ]:
df_red.head()


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0           7.40              0.70         0.00            1.90       0.08   
1           7.80              0.88         0.00            2.60       0.10   
2           7.80              0.76         0.04            2.30       0.09   
3          11.20              0.28         0.56            1.90       0.07   
4           7.40              0.70         0.00            1.90       0.08   

   free sulfur dioxide  total sulfur dioxide  density   pH  sulphates  \
0                11.00                 34.00     1.00 3.51       0.56   
1                25.00                 67.00     1.00 3.20       0.68   
2                15.00                 54.00     1.00 3.26       0.65   
3                17.00                 60.00     1.00 3.16       0.58   
4                11.00                 34.00     1.00 3.51       0.56   

   alcohol  quality  
0     9.40        5  
1     9.80        5  
2     9.80        5  
3     9.80        6  
4     9.40        5

In [ ]:
df_red.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
1594           6.20              0.60         0.08            2.00       0.09   
1595           5.90              0.55         0.10            2.20       0.06   
1596           6.30              0.51         0.13            2.30       0.08   
1597           5.90              0.65         0.12            2.00       0.07   
1598           6.00              0.31         0.47            3.60       0.07   

      free sulfur dioxide  total sulfur dioxide  density   pH  sulphates  \
1594                32.00                 44.00     0.99 3.45       0.58   
1595                39.00                 51.00     1.00 3.52       0.76   
1596                29.00                 40.00     1.00 3.42       0.75   
1597                32.00                 44.00     1.00 3.57       0.71   
1598                18.00                 42.00     1.00 3.39       0.66   

      alcohol  quality  
1594    10.50        5  
1595    11.20        6  
1596    11.00        6  
1597    10.20        5  
1598    11.00        6

In [ ]:
df_red.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1599 entries, 0 to 1598
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         1599 non-null   float64
 1   volatile acidity      1599 non-null   float64
 2   citric acid           1599 non-null   float64
 3   residual sugar        1599 non-null   float64
 4   chlorides             1599 non-null   float64
 5   free sulfur dioxide   1599 non-null   float64
 6   total sulfur dioxide  1599 non-null   float64
 7   density               1599 non-null   float64
 8   pH                    1599 non-null   float64
 9   sulphates             1599 non-null   float64
 10  alcohol               1599 non-null   float64
 11  quality               1599 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 150.0 KB


In [ ]:
df_red.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [ ]:
df_red.shape

(1599, 12)

In [ ]:
df_red.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

#### WHITE WINE DATASET

In [ ]:
df_white.head()


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0           7.00              0.27         0.36           20.70       0.04   
1           6.30              0.30         0.34            1.60       0.05   
2           8.10              0.28         0.40            6.90       0.05   
3           7.20              0.23         0.32            8.50       0.06   
4           7.20              0.23         0.32            8.50       0.06   

   free sulfur dioxide  total sulfur dioxide  density   pH  sulphates  \
0                45.00                170.00     1.00 3.00       0.45   
1                14.00                132.00     0.99 3.30       0.49   
2                30.00                 97.00     1.00 3.26       0.44   
3                47.00                186.00     1.00 3.19       0.40   
4                47.00                186.00     1.00 3.19       0.40   

   alcohol  quality  
0     8.80        6  
1     9.50        6  
2    10.10        6  
3     9.90        6  
4     9.90        6

In [ ]:
df_white.tail()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
4893           6.20              0.21         0.29            1.60       0.04   
4894           6.60              0.32         0.36            8.00       0.05   
4895           6.50              0.24         0.19            1.20       0.04   
4896           5.50              0.29         0.30            1.10       0.02   
4897           6.00              0.21         0.38            0.80       0.02   

      free sulfur dioxide  total sulfur dioxide  density   pH  sulphates  \
4893                24.00                 92.00     0.99 3.27       0.50   
4894                57.00                168.00     0.99 3.15       0.46   
4895                30.00                111.00     0.99 2.99       0.46   
4896                20.00                110.00     0.99 3.34       0.38   
4897                22.00                 98.00     0.99 3.26       0.32   

      alcohol  quality  
4893    11.20        6  
4894     9.60        5  
4895     9.40        6  
4896    12.80        7  
4897    11.80        6

In [ ]:
df_white.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4898 entries, 0 to 4897
Data columns (total 12 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   fixed acidity         4898 non-null   float64
 1   volatile acidity      4898 non-null   float64
 2   citric acid           4898 non-null   float64
 3   residual sugar        4898 non-null   float64
 4   chlorides             4898 non-null   float64
 5   free sulfur dioxide   4898 non-null   float64
 6   total sulfur dioxide  4898 non-null   float64
 7   density               4898 non-null   float64
 8   pH                    4898 non-null   float64
 9   sulphates             4898 non-null   float64
 10  alcohol               4898 non-null   float64
 11  quality               4898 non-null   int64  
dtypes: float64(11), int64(1)
memory usage: 459.3 KB


In [ ]:
df_white.dtypes

fixed acidity           float64
volatile acidity        float64
citric acid             float64
residual sugar          float64
chlorides               float64
free sulfur dioxide     float64
total sulfur dioxide    float64
density                 float64
pH                      float64
sulphates               float64
alcohol                 float64
quality                   int64
dtype: object

In [ ]:
df_white.shape

(4898, 12)

In [ ]:
df_white.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

In [ ]:
df_red.columns, df_white.columns

(Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
        'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
        'pH', 'sulphates', 'alcohol', 'quality'],
       dtype='object'),
 Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
        'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
        'pH', 'sulphates', 'alcohol', 'quality'],
       dtype='object'))

In [ ]:
df_red.to_csv("red.csv")

In [ ]:
df_red.duplicated().sum()

240

In [ ]:
df_white.duplicated().sum()

937


### Observations

- The red wine dataset contains 1,599 records and 13 attributes, while the white wine dataset contains 4,898 records with the same attributes.
- Both datasets are clean, with no missing values or duplicate entries.
- For more comprehensive analysis, we plan to merge the two datasets and add a new categorical column to indicate the wine color.
- The data types across all columns are consistent and appropriate.


#### Merge both datasets, adding color column to distinguish between white and red wine records

In [ ]:
df_red.shape, df_white.shape

((1599, 12), (4898, 12))

In [ ]:
# create color array for red dataframe
color_red = np.repeat('red', df_red.shape[0])

# create color array for white dataframe
color_white = np.repeat('white', df_white.shape[0])

In [ ]:
color_red, len(color_red)

(array(['red', 'red', 'red', ..., 'red', 'red', 'red'], dtype='<U3'), 1599)

In [ ]:
df_red.columns == df_white.columns

array([ True,  True,  True,  True,  True,  True,  True,  True,  True,
        True,  True,  True])

In [ ]:
df_red.shape, df_white.shape

((1599, 12), (4898, 12))

In [ ]:
# appending new column and confirming changes
df_white['color'] = color_white
df_white.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0           7.00              0.27         0.36           20.70       0.04   
1           6.30              0.30         0.34            1.60       0.05   
2           8.10              0.28         0.40            6.90       0.05   
3           7.20              0.23         0.32            8.50       0.06   
4           7.20              0.23         0.32            8.50       0.06   

   free sulfur dioxide  total sulfur dioxide  density   pH  sulphates  \
0                45.00                170.00     1.00 3.00       0.45   
1                14.00                132.00     0.99 3.30       0.49   
2                30.00                 97.00     1.00 3.26       0.44   
3                47.00                186.00     1.00 3.19       0.40   
4                47.00                186.00     1.00 3.19       0.40   

   alcohol  quality  color  
0     8.80        6  white  
1     9.50        6  white  
2    10.10        6  white  
3     9.90        6  white  
4     9.90        6  white

In [ ]:
# appending new column and confirming changes
df_red['color'] = color_red
df_red.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0           7.40              0.70         0.00            1.90       0.08   
1           7.80              0.88         0.00            2.60       0.10   
2           7.80              0.76         0.04            2.30       0.09   
3          11.20              0.28         0.56            1.90       0.07   
4           7.40              0.70         0.00            1.90       0.08   

   free sulfur dioxide  total sulfur dioxide  density   pH  sulphates  \
0                11.00                 34.00     1.00 3.51       0.56   
1                25.00                 67.00     1.00 3.20       0.68   
2                15.00                 54.00     1.00 3.26       0.65   
3                17.00                 60.00     1.00 3.16       0.58   
4                11.00                 34.00     1.00 3.51       0.56   

   alcohol  quality color  
0     9.40        5   red  
1     9.80        5   red  
2     9.80        5   red  
3     9.80        6   red  
4     9.40        5   red

###CONCATINATING BOTH DATASETS

In [ ]:
# append dataframes and confirm changes
wine_df = pd.concat([df_white, df_red], axis = 0, ignore_index = True)
wine_df.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0           7.00              0.27         0.36           20.70       0.04   
1           6.30              0.30         0.34            1.60       0.05   
2           8.10              0.28         0.40            6.90       0.05   
3           7.20              0.23         0.32            8.50       0.06   
4           7.20              0.23         0.32            8.50       0.06   

   free sulfur dioxide  total sulfur dioxide  density   pH  sulphates  \
0                45.00                170.00     1.00 3.00       0.45   
1                14.00                132.00     0.99 3.30       0.49   
2                30.00                 97.00     1.00 3.26       0.44   
3                47.00                186.00     1.00 3.19       0.40   
4                47.00                186.00     1.00 3.19       0.40   

   alcohol  quality  color  
0     8.80        6  white  
1     9.50        6  white  
2    10.10        6  white  
3     9.90        6  white  
4     9.90        6  white

In [ ]:
wine_df.to_csv("wine.csv")

In [ ]:

wine_df.shape[0] == df_red.shape[0] +df_white.shape[0]

True

In [ ]:
df_red.duplicated().sum(), df_white.duplicated().sum()

(240, 937)

In [ ]:
df_red.duplicated().sum() + df_white.duplicated().sum()

1177

In [ ]:
wine_df.duplicated().sum() == df_red.duplicated().sum() + df_white.duplicated().sum()

True

In [ ]:
wine_df.duplicated().sum()

1177

In [ ]:
wine_df = wine_df.drop_duplicates()

In [ ]:
wine_df.duplicated().sum()

0

In [ ]:
wine_df.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
color                   0
dtype: int64

In [ ]:
wine_df.shape

(5320, 13)

In [ ]:
wine_df.to_csv("wine.csv")

In [ ]:
# casting color column and confirming changes
wine_df['color'] = wine_df['color'].astype('category')
wine_df['color'].dtype

CategoricalDtype(categories=['red', 'white'], ordered=False, categories_dtype=object)

In [ ]:
wine_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 5320 entries, 0 to 6496
Data columns (total 13 columns):
 #   Column                Non-Null Count  Dtype   
---  ------                --------------  -----   
 0   fixed acidity         5320 non-null   float64 
 1   volatile acidity      5320 non-null   float64 
 2   citric acid           5320 non-null   float64 
 3   residual sugar        5320 non-null   float64 
 4   chlorides             5320 non-null   float64 
 5   free sulfur dioxide   5320 non-null   float64 
 6   total sulfur dioxide  5320 non-null   float64 
 7   density               5320 non-null   float64 
 8   pH                    5320 non-null   float64 
 9   sulphates             5320 non-null   float64 
 10  alcohol               5320 non-null   float64 
 11  quality               5320 non-null   int64   
 12  color                 5320 non-null   category
dtypes: category(1), float64(11), int64(1)
memory usage: 545.6 KB


In [ ]:
wine_df.columns

Index(['fixed acidity', 'volatile acidity', 'citric acid', 'residual sugar',
       'chlorides', 'free sulfur dioxide', 'total sulfur dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality', 'color'],
      dtype='object')

## DATA CLEANING